xcommon/src/llvm.rs

```diff
@@ -0,0 +1,121 @@
+//! LLVM utilities
+
+use anyhow::{bail, ensure, Context, Result};
+use std::collections::HashSet;
+use std::path::{Path, PathBuf};
+use std::process::Command;
+
+/// Returns the set of additional libraries that need to be bundled with
+/// the given library, scanned recursively.
+///
+/// Any libraries in `provided_libs_paths` will be treated as available, without
+/// being emitted. Any other library not in `search_paths` or `provided_libs_paths`
+/// will result in an error.
+pub fn list_needed_libs_recursively(
+    lib: &Path,
+    search_paths: &[&Path],
+    provided_libs_paths: &[&Path],
```

Comment:
That sounds like a lot of work to update and maintain?

Reply:
Also if we don't update this with new libraries, the logic will just copy the new libraries even when they are already available on the phone, making an APK unnecessarily bigger than it needs to be.

Reply:
Note that none of the items in `provided_libs_paths` is used to copy or blacklist anything at all. That only happens for `search_paths`. The goal of `provided_libs_paths` is to at least validate that a library is available (without copying it!) so that we can otherwise provide an error instead of having the app fail at runtime.

Reply:
for linux we already have an excludelist [0]. for android we can compute the exclude list by finding all libraries in provided_libs_path - libc++_shared.so. if it's in the exclude list we skip it otherwise we use search_paths to locate it and include it. for macos/windows we can compute the exclude list similarly.

- [0] https://github.com/AppImage/pkg2appimage/blob/master/excludelist

Reply:
In this sense `provided_libs_paths` already is a blocklist, we just need to remove `libc++_shared.so`. Are you asking for this list to be checked in to the repo somewhere, or simply to refactor this so that it's still computed at runtime but more clear in the code? I can do the latter :)

Reply:
well, it contains the library name not the path. the block list would be all file names with an so extension within provided paths. so something like `filename(find search_paths | grep *.so)`

Reply:
in a sense it's already an excludelist, was just  suggesting how to generalize it. it requries a bit more work but should be more generic.

Reply:
I think you have a fair point yes, but as pointed out below perhaps you weren't aware that `provided_libs_paths` is *not* included in `search_paths` currently.

Perhaps we can improve that as I quite like your suggestion. It is indeed already converted into an exclude list, with `libc++_shared.so` removed from it:

https://github.com/cloudpeers/xbuild/blob/30b0849d20b22825b233d19245a3f1c808e787dc/xcommon/src/llvm.rs#L22-L29

Reply:
> ah, this is good enough for now, let's just merge it.

Your call, we can also merge it now and improve in the future. I don't mind either way, I think we just misunderstood each others ideas in the beginning :)